### ISSUE TITLE
Add code for 'unpacking' FFTs

### ISSUE THREAD

**Issue by astewart-xmos** [xmos/lib_xs3_math#38]
**TLDR:**  Add functions for unpacking/repacking the Nyquist component of the DFT so that the spectrum can be used in a simpler and more natural way.

---

In order to compute the DFT in-place, both `bfp_fft_forward_mono()` and `bfp_fft_forward_stereo()` functions both take advantage of the complex conjugate symmetry of the DFT of a real signal.  This property, along with the periodicity of the DFT implies that the DC and Nyquist components of the output spectrum must have zero imaginary part.  That being the case, the DFT can be computed in-place by packing the real part of the Nyquist component of the spectrum into the imaginary part of the DC component.

Sometimes this packing is fine as-is. For example, if the spectrum is to be added to another (similarly-packed) spectrum, or the spectrum is element-wise multiplied by a vector of reals. However, certain other operations result in (almost certainly) undesirable results, such as computing the squared magnitude (i.e. energy/power) of spectral components, in which case we end up with a single real value capturing a meaningless combination of the DC and Nyquist components.

One way of dealing with this (as has been done in e.g. `lib_aec`) is to extract the Nyquist component from the spectrum entirely, and simply deal with it separately.  Unfortunately, this results in a lot of redundant-looking code, potential confusion when dealing with headroom and exponents, and a possibly-increased run-time (relative the the proposed solution which follows).

An alternative solution is to just unpack the spectra into a larger buffer, so that the spectrum of an e.g. 512-point FFT has 257 components [0, 256] instead of 256 components [0, 256).  Once unpacked, operations can be applied to spectrum in a straight-forward way, avoiding most of the down side of the other solution.

For a mono FFT this solution is practically free. It requires allocating two extra elements (`int32_t`) to the input signal's buffer, and then modifying the output vector by adding 1 to its length and moving the Nyquist component to its proper location.

For a stereo FFT this solution incurs a slightly larger cost. It also requires allocating two extra elements (this time `ch_pair_s32_t`) to the input buffer. However, this time, the spectrum for Channel B must be shifted over one element in memory (two a higher address), which of course is linear in the FFT length. Then the two spectra can be unpacked as with the mono FFT.  (Note that in the case of something like AEC which requires many FFTs/IFFTs per audio frame the cost multiplies.)

Of course, in both the mono and stereo cases, the process must be inverted before an IFFT can be performed.

Because the process of unpacking and repacking the mono and stereo DFT spectra requires multiple steps undertaken in the correct order, it is an easy place to introduce bugs.  Currently, `lib_xs3_math` offers no support for this technique.

There are two obvious options to support this:
* Add callable mono/stereo unpack/repack functions
* Add a flag to the BFP FFT functions which indicates whether the user would like the spectra to be unpacked/repacked.

For now I suggest the former.  It would be straight-forward at a later time to add the flags to the FFT functions and just have them call the stand-alone unpack/repack functions.

**Comment by astewart-xmos** [comment]
* The FFT example app should also be updated to demonstrate these.

**Comment by astewart-xmos** [comment]
Added these functions.  I didn't add a flag to the forward/inverse FFT functions.

It would be nice if there was a way to verify that the user allocated enough space in the underlying mantissa buffer to accommodate the unpacking as that may be a source of error, but I know of a good way to do that.
